pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go

```diff
@@ -21,68 +20,74 @@ type AzureTokenProvider interface {
+func NewAzureAccessTokenProvider(cfg *setting.Cfg, credentials azcredentials.AzureCredentials, scopes []string) *tokenProviderImpl {
 	return &tokenProviderImpl{
-		cfg:        cfg,
-		authParams: authParams,
+		cfg:         cfg,
+		credentials: credentials,
+		scopes:      scopes,
 	}
 }
 
 func (provider *tokenProviderImpl) GetAccessToken(ctx context.Context) (string, error) {
 	var credential TokenCredential
 
-	if provider.isManagedIdentityCredential() {
+	// TODO: Move to provider initialization and reuse for each GetAccessToken call
+	switch c := provider.credentials.(type) {
+	case *azcredentials.AzureManagedIdentityCredentials:
 		if !provider.cfg.Azure.ManagedIdentityEnabled {
 			err := fmt.Errorf("managed identity authentication is not enabled in Grafana config")
 			return "", err
 		} else {
-			credential = provider.getManagedIdentityCredential()
+			credential = provider.getManagedIdentityCredential(c)
 		}
-	} else {
-		credential = provider.getClientSecretCredential()
+	case *azcredentials.AzureClientSecretCredentials:
+		credential = provider.getClientSecretCredential(c)
+	default:
+		err := fmt.Errorf("credentials of type '%s' not supported by authentication provider", c.AzureAuthType())
+		return "", err
 	}
 
-	accessToken, err := azureTokenCache.GetAccessToken(ctx, credential, provider.authParams.Scopes)
+	accessToken, err := azureTokenCache.GetAccessToken(ctx, credential, provider.scopes)
 	if err != nil {
 		return "", err
 	}
 
 	return accessToken, nil
 }
 
-func (provider *tokenProviderImpl) isManagedIdentityCredential() bool {
-	authType := strings.ToLower(provider.authParams.Params["azure_auth_type"])
-	clientId := provider.authParams.Params["client_id"]
-
-	// Type of authentication being determined by the following logic:
-	// * If authType is set to 'msi' then user explicitly selected the managed identity authentication
-	// * If authType isn't set but other fields are configured then it's a datasource which was configured
-	//   before managed identities where introduced, therefore use client secret authentication
-	// * If authType and other fields aren't set then it means the datasource never been configured
-	//   and managed identity is the default authentication choice as long as managed identities are enabled
-	return authType == "msi" || (authType == "" && clientId == "" && provider.cfg.Azure.ManagedIdentityEnabled)
-}
-
-func (provider *tokenProviderImpl) getManagedIdentityCredential() TokenCredential {
-	clientId := provider.cfg.Azure.ManagedIdentityClientId
-
-	return &managedIdentityCredential{clientId: clientId}
+func (provider *tokenProviderImpl) getManagedIdentityCredential(credentials *azcredentials.AzureManagedIdentityCredentials) TokenCredential {
+	var clientId string
+	if credentials.ClientId != "" {
+		clientId = credentials.ClientId
+	} else {
+		clientId = provider.cfg.Azure.ManagedIdentityClientId
+	}
+	return &managedIdentityCredential{
```

Comment:
also unrelated to this PR but why having an `Init` method as part of the interface if we could do that in this method? (reducing the risk of calling `GetAccessToken` without the struct being initialized.)

Reply:
`TokenCredential` struct serves two purposes:
- as a cache key accessor
- as an abstraction layer over the GetAccessToken implementation (Azure SDK's `azidentity`).

In the old implementation of token provider `TokenCredential` was created on every call of `GetAccessToken()`  and in majority of cases was thrown away unused after checking the key because `ConcurrentTokenCache` always use the instance of `TokenCredential` which it has already in cache.

`Init` method has `azidentity` initialization which is expensive, for example `azidentity.NewManagedIdentityCredential` does probing of endpoints to determine managed identities implementation. So I didn't want to have expensive initialization only to throw out the struct. Only initialization of the struct fields is cheap.

`ConcurrentTokenCache` will call `Init` only in case if it actually intends to use it.

In the new implementation I have moved the `TokenCredential` creation to constructor to make implementation of `AzureTokenProvider` even more lightweight. In this case `Init` is useful to make sure that `azidentity` being initialized during the `GetAccessToken` call, not during construction of `AzureTokenProvider`.   Init can fail, and I want it to fail during GetAccessToken, to properly propagate the error as result of `GetAccessToken`.